### DISCUSSION TITLE
Making the dream Animation API for GDExtension

### DISCUSSION THREAD

**Discussion by GuilhermeGSousa** [GuilhermeGSousa/godot-motion-matching#52] [Ideas]
So right now, there are no good ways to making motion matching as a GDExtension. There are some features missing that prevent us from extending certain engine functionalities. In this post I'll go into detail of what would be required on the engine side to do this. For each point I'll go over what we're trying to do and what the ideal engine API would look like to be able to support it.

### What we want to do
From implementing it in an engine module, its has become clear that the best way of implementing motion matching in Godot is as an `AnimationNode`. Doing so comes with multiple advantages compared to other approaches we've tried:
- Blend the result of motion matching with other animations
- Use state machines to switch between motion matching and regular animation techniques
- Run motion matching queries on smaller animation libraries, and switch between them using state machines.

#### The `_process` method, arguments and return values
Probably the most important part of this section, to be able to extend a `AnimationRootNode`, both `AnimationMixer::PlaybackInfo` and `AnimationNode::NodeTimeInfo` need to be ref counted, so that the following function can have a `GDVIRTUAL` equivalent:

```c++
NodeTimeInfo _process(const PlaybackInfo p_playback_info, bool p_test_only = false);
```

#### Accessing the `AnimationTree`

**During `_process`**, `AnimationNode::get_animation_tree()` can be used to access the `AnimationTree` that owns the node. We'd need this same function exposed to extensions to be able to create a Motion Matching animation node.

**At editor time (outside the `_process` function)**, that same method will always return null. In order to be able to implement a similar property validator as the one used by `AnimationNodeAnimation`, we'd need a way of accessing the animation tree from `_validate_property`. From a module, this can be done the following way:
```c++
#ifdef TOOLS_ENABLED
AnimationTree* tree = AnimationTreeEditor::get_singleton()->get_animation_tree();
#endif
```
We'd need that same API exposed to gd extensions as well.

### What we want to do
Inertialization blending is a technique used to blend animations together, [developed by the teams at Microsoft Studios for Gears of War 4](https://www.youtube.com/watch?v=BYyv4KTegJI). One of its characteristics is that the rate of change of the bones preserve some of their inertia when swapping animations, resulting in more fluid and realistic motion. It can also be more performant than "regular" track blending, as it is not required to evaluate two tracks simultaneously to compute the resulting blend.

Since this blending technique requires modifying the skeleton bone's position and rotations, implementing this as a `SkeletonModifier3D` is ideal. However, there are still a couple of missing pieces to the puzzle.

#### Signals from `AnimationNode`s
The inertialization blending algorithm needs to run specific calculations when animations change (i.e. when motion matching run a query and selects an animation). We therefore need a way for animation nodes to be able to signal outside systems when something happens. An idea of how to do it would be to have a similar method as `AnimationNode::get_parameter_list` that would expose signals instead:

```
Array[Dictionary] get_signal_list() const
```
(similar to the method with the same name of `Object`).

**Comment by fire** [comment]
Upcoming work that'll allow compiling godot-cpp gdextensions as a module. 💯

**Reply by fire** [reply]
@TokageItLab What do you think? I'm not too fond of the signals approach, but maybe we can use call-backs or something else.

**Reply by lyuma** [reply]
I agree with fire's comment. Signals have too many problems, including thread safety and performance, to be used for this purpose.

> Since this blending technique requires modifying the skeleton bone's position and rotations, implementing this as a SkeletonModifier3D is ideal

I agree with SkeletonModifier3D, and I'm confused what precisely is missing from here.
I think I have an idea: instead of using signals, perhaps it could be possible to query the currently blended animations, or query if a blend is currently happening.

Is it already possible to query the blending by using get_parameter() to get the blend values? but having a more efficient way to do this would be nice.

**Reply by GuilhermeGSousa** [reply]
That could work too, and then have whatever system needs to know when a new blend started pool that information on process? The motion matching node could set a parameter to true whenever a motion matching query happens, for the inertialization skeleton modifier to read, but I would prefer some event based solution as opposed to having to pooling of a variable.

I guess TL;DR, I'm looking for a way for events that happen inside an `AnimationRootNode`'s process function to notify other systems.

**Reply by GuilhermeGSousa** [reply]
>  query the currently blended animations, or query if a blend is currently happening.

That said I could probably make things works by having either of these

**Reply by TokageItLab** [reply]
In conclusion, I believe it makes more sense to create a class for blending SkeletonModifier results than to implement IK, MotionMatching results, etc. as AnimationNodes for blending.

Since the blending API has considerable influence on the other AnimationTracks, it is unsafe to interrupt the blending API in AnimationMixer. There is already an API with a similar role, post_process_key_value(), but it is not sufficient to handle motion matching.

As I remember from the discussion on mirroring animation results, what is needed for this kind of API is something like an AnimationMixerResultMixer that further blends the results of the AnimationTree.

If we ignore AudioTrack, ValueTrack and other tracks except TRS3DTracks, this can be accomplished to some extent by implementing a SkeletonModifier with its own blending API. So for example, we can make AnimationResultStacker which extends SkeletonModifier that when there are multiple AnimationMixers, the results of the bone pose within the mixer_applied signal can be temporarily cached to stack the application timings.

The remaining problem is that SkeletonModifier can only perform serial blending. As mentioned in previous discussions, two SkeletonModifiers that refer to the same pose cannot exist at the same time. This is because a SkeletonModifier can only refer to a pose after the previous SkeletonModifier's processing has been applied.

It is possible to get a bone_pose from each individual connection to the modification_processed signal to pose at an arbitrary time, but I doubt that all SkeletonModifiers will do so.

One idea that comes to mind is to create a class called SkeletonModifierMixer which extends SkeletonModifier that is assumed to be placed at the end of the SkeletonModifier list. It would connect to the modification_processed signal of all SkeletonModifiers before itself in the list, stack all the bone poses they have applied, and finally blend them again and apply them to the Skeleton. Also, after caching the pose of the modification_processed signal, the effect of the corresponding SkeletonModifier is reset there, if necessary; For example, if the process sequence is AnimationResult -> IK1 -> IK2 -> IK3 -> SkeletonModifierMixer, SkeletonModifierMixer can reset the results of IK1 and IK2 in the modification_processed IK3 will receive the result of AnimationResult. --Although it may be necessary to create a slightly larger GUI like AnimationTreeEditor in order to make the interface easy to use.

**Reply by fire** [reply]
I would like there to be a way to use this as a AnimationNode and as a SkeletonModifier as both have their own uses.

Is that possible?

**Reply by TokageItLab** [reply]
For reasons of process order issues, I believe that only things that can be implemented in AnimationNode that do not depend on the results of other Nodes can be implemented in AnimationNode, so I do not think it makes much sense to have IK, MotiomMatching and other SkeletonModifications in AnimationNode.

For example, consider what you have for MotionMatching estimation. If the current position of the character is used, MotionMatching must be processed after the Animation is applied if the position is changed by the Animation. In other words, MotionMatching cannot be placed in parallel on top of the AnimationMixer that processes the Animation.

Conversely, let's think about what an AnimationNode can do. The answer is such as an implementation that plays its own animation format, such as Unity's MuscleAnimation or ShaderMotion. This is because it does not depend on the state of other Nodes, but only on the conversion of data independent of the Node through static parameters.

In other words, MotionMatching can be an AnimationNode as long as the values that are the source of MotionMatching estimation are not changed by Animation, but I don't think that is the right way to go.

As I understand it, the source of MotionMatching is the physical change in bone orientation due to IK and collisions, but that change is done by the movement of the character. It is very likely that the movement of the character is done by animation using root motion, so if MotionMatching's restrictions do not allow the use of root motion, it is obviously useless.

Also, given that MotionMatching needs to be processed after IK in order to source IK values, if MotionMatching is an AnimationNode, then MotionMatching must have its own IK or Skeleton within it internally. If MotionMatching is a SkeletonModifier, there is no need for such a redundant implementation, and the user can use whatever IK settings they wish.

**Reply by fire** [reply]
> Values that are the source of MotionMatching estimation are not changed by Animation

According to this chart, the features are precomputed, so the animation does not change them.

The trajectory is input given by the animation system. For example, as input, it does not vary based on previous frames. 

[Inertialization](https://theorangeduck.com/page/inertialization-transition-cost) should be a SkeletonModifier3D though.

Did I get this correct?

```mermaid
flowchart LR
    %% Frames Row
    subgraph Frames [Frames]
        direction LR
        F1["Frame 1"]
        F2["Frame 2"]
        F3["Frame 3"]
    end

    %% Feature Blocks for Each Frame
    subgraph FeatureBlock1 [Feature Block for Frame 1]
        direction TB
        A1["Feature 1: Root past/future trajectory (3-5 floats per point)"]
        B1["Feature 2: Left foot position (3 floats)"]
        C1["Feature 3: Left foot velocity (3 floats)"]
    end

    subgraph FeatureBlock2 [Feature Block for Frame 2]
        direction TB
        A2["Feature 1: Root past/future trajectory (3-5 floats per point)"]
        B2["Feature 2: Left foot position (3 floats)"]
        C2["Feature 3: Left foot velocity (3 floats)"]
    end

    subgraph FeatureBlock3 [Feature Block for Frame 3]
        direction TB
        A3["Feature 1: Root past/future trajectory (3-5 floats per point)"]
        B3["Feature 2: Left foot position (3 floats)"]
        C3["Feature 3: Left foot velocity (3 floats)"]
    end

    %% Independent Connections
    F1 --> FeatureBlock1
    F2 --> FeatureBlock2
    F3 --> FeatureBlock3
```

**Reply by TokageItLab** [reply]
I think the major problem with that graph is that it uses the values from the previous frame, meaning that the correct calculation is not done on the first frame of the user input.

What is the source of the value in the second frame? For example, if the input is initial on the first frame, even if the result of MotionMatching plays idle motion, the movement required to walk on the second frame is the movement of the character, and the source of the character's movement could well be an animation outside of MotionMatching system.

However, if MotionMatching is an AnimationNode and the animation is applied from a parallel AnimationMixer, MotionMatching will actually not be able to detect movement until the next frame. If MotionMatching is a SkeletonModifier, MotionMatching can detect a walk immediately after the character is moved by Animation without waiting for the next frame.

**Reply by GuilhermeGSousa** [reply]
Just to clarify some things, all that motion matching needs to do is change animations based on parameters provided to the animation tree. In a sense, it is not much different from a state machine, blending from one animation to the next based on some state variables. In this sense, motion matching is completely self contained and requires no other input from other animation node, which is why I believe it works well as an animation node. It also does not require the results of IK: IK should be done as a "cleanup" pass after motion matching has run.

Since the input provided to the motion matching animation node already contains all the state information it needs to run motion matching queries (state that includes the past and future trajectory points, the current positions/rotations/velocities of bones, and everything else we might need), it would even be possible for motion matching queries to run asynchronously on another thread. The only downside being that it could be running its logic on information that is one or two frames old, which from experience shouldn't be noticeable and is to be expected when running async codee.

> It is very likely that the movement of the character is done by animation using root motion, so if MotionMatching's restrictions do not allow the use of root motion, it is obviously useless.

Since motion matching only needs to select and play animations, movement is still done using root motion in our implementation. Users can then choose to either use root motion or code based movement logic, but that is something else entirely.

That are also issues that arise from motion matching not being an animation node: motion matching is often used for locomotion animations, and we often want to blend those animation with others (like being able to play an "aiming" animation with the locomotion from motion matching. This can be done with no extra effort with blend nodes and bone filters in the animation tree, but is much harder to do if motion matching was a SkeletonModifier3D.

> One idea that comes to mind is to create a class called SkeletonModifierMixer which extends SkeletonModifier that is assumed to be placed at the end of the SkeletonModifier list.

I like this idea a lot! Having multiple IK running for a same bone is often done in AAA games, and being able to a specify an "influence" weight for each SkeletonModifier3D would be a very welcome feature.

**Reply by GuilhermeGSousa** [reply]
@TokageItLab we have a working demo of the motion matching animation node on our engine-module branch if you ever wish to take a look!
https://github.com/GuilhermeGSousa/godot-motion-matching/blob/engine-module/src/mm_animation_node.cpp

**Reply by TokageItLab** [reply]
My concern is that the query you mention could be input by animation. You mean that queries such as character_transform cannot be input by animation or external IK, is that common in MotionMatching?

There is definitely a concern about blending with root motion in SkeletonModifier, but I think this can be resolved by processing the https://github.com/godotengine/godot/pull/90361#issuecomment-2309978354 comment with SkeletonModifierMixer or something similar.

**Reply by GuilhermeGSousa** [reply]
I can only really speak to the experience I've had implementing motion matching at work (in Unreal), but we don't require any results from IK to run the motion matching query. `character_transform` will be the transform of the skeleton at the time of the query. Depending on the use case, that transform can either be set by root motion, or by other systems: at Naughty Dog for example, for The Last Of Us 2 they move the skeleton using root motion for NPCs, but move it using movement code for the player. I highly recommend their GDC talk btw, it has helped us a lot to setup our own motion matching implementation.
https://gdcvault.com/play/1027378/Motion-Matching-in-The-Last

Either way, there's no hard requirement as to where `character_transform` is for motion matching to work: it could start at a random location on a plane, and through root motion of animations selected by motion matching queries it would eventually reach the character's capsule (i.e. the CharacterBody3D). This is something we can demonstrate with our current demo.

**Reply by GuilhermeGSousa** [reply]
Something to note, given a node tree of a character using motion matching:
```
- MMCharacter (CharacterBody3D)
    - Skeleton3D
    - AnimationPlayer
    - AnimationTree
```
One peculiarity of this system is that Skeleton3D will be set as "top level" so that it and MMCharacter can move independently. This is done so that users can be the ones to set how the character should move. Going back to Naughty Dog's example:
- Option 1: Either MMCharacter moves with Skeleton3D (100% root motion, done for NPCs)
- Option 2: Skeleton3D moves with MMCharacter (100% code based movement, done for the player since it usually has better controllability)
- Option 3: A mix of both (I think Epic is currently implementing this for their upcoming release of Motion Matching for Unreal)

**Reply by TokageItLab** [reply]
Then, there are problems in Option 1 and 3.

One is the one frame delay problem, as mentioned above.

And the more serious problem is that if there are extra AnimationMixers, they can completely overwrite the MotionMatching results, depending on the order of processing in the SceneTree. This is the main reason why SkeletonModifier was implemented and why IK should not be an AnimationNode.

**Reply by GuilhermeGSousa** [reply]
I 100% agree with you @TokageItLab on SkeletonModifiers, I believe all IK should happen on SkeletonModifiers, which is why the Motion Matching animation node does not do any IK. It acts similarly to the `AnimationNodeAnimation`, except the animations it plays are coming from a search on a database. All IK should happen as a "post-processing" pass that happens after motion matching.

**Reply by GuilhermeGSousa** [reply]
As for extra Animation Mixers, I must admit I'm not familiar with that setup, could you tell me more about how that works? I could imagine other animation nodes overwriting motion matching (e.g. play a different animation for all the torso bones), but that would 
be happening in the same animation mixer right?

**Reply by TokageItLab** [reply]
No. If two AnimationMixers are playing at the same frame, the AnimationMixer that plays later on the SceneTree always takes precedence. This is true even if the MotionMatching is an independent node; only the bone pose processed later will always have precedence.

In your setup above, if you move MMCharacter using Option2 by script independent from AnimationMixer, depending on the process order between script and AnimationMixer, either the result of the script movement or the result of MotionMatching will be completely ignored. Since the AnimationPlayer can be placed anywhere, it is not in line with Godot's design to force the placement of the AnimationPlayer on the MMCharacter's children.

If MotionMatching is an AnimationNode, moving the character in the same AnimationMixer will delay the process by one frame. However, if extra AnimationMixer is used to move the character, only one of the results will always have precedence and cannot blending. It cannot be said that the correct calculations are made at the correct time in either case.

SkeletonModifier solves this kind of order-dependent problem by having the SkeletonModifier process independently of the AnimationMixer order in the scene tree. It always processes the list of SkeletonModifiers, which are children of Skeleton, from the top, after AnimationMixer is processed. This is why we suggested above the AnimationMixerStacker, etc. See also https://godotengine.org/article/design-of-the-skeleton-modifier-3d/.

**Reply by GuilhermeGSousa** [reply]
In that case, wouldn't the frame dependent movement issue be solved by having root motion be applied by a SkeletonModifier? So something like this:

1. `MMCharacter` makes all the necessary calculations to generate a future and past trajectory, and generates a `MMQueryInput`
2. `MMCharacter` calls `animation_tree->set("parameters/motion_matching_input", query_input)`, thus setting providing the necessary data for motion matching selections
3. `MMAnimationNode` call `get_parameter` to retrieve the `MMQueryInput`, makes a query resulting in a new animation to be played, and plays that animation.
4. A skeleton modifier calls the following code to move the skeleton using root motion:
```c++
void _apply_root_motion() {
    _skeleton->set_quaternion(
        _skeleton->get_quaternion() * _animation_tree->get_root_motion_rotation());

    const Vector3 movement_delta = (_animation_tree->get_root_motion_rotation_accumulator().inverse() * _skeleton->get_quaternion()).xform(_animation_tree->get_root_motion_position());

    _skeleton->set_global_position(_skeleton->get_global_position() + movement_delta);
}
```
5. Some other skeleton modifier makes the necessary calculations for IK

To be clear, I also don't think movement should be done in the motion matching animation node, all that should do is play an animation. Our current demo handles root motion outside of the animation mixer (which doesn't necessarily mean it isn't delayed by a frame).

**Reply by TokageItLab** [reply]
Thinking back to fundamentals, it seems to me that the problem lies in MotionMatching's extension of AnimationNode. The most problematic issue here is the possibility of MotionMatching's next query input being played back in parallel, which must be avoided.

So probably the cleanest way forward is for MotionMatching to extend AnimationMixer directly, rather than AnimationNode. Then, blending (with other animations outside of MotionMatching) should be implemented as a AnimationMixerStacker extends SkeletonModifier, and delegated to it.

To get the delta of bones modified by IK etc., MotionMatching would probably need to get the final pose of the skeleton with the Skeleton3d.skeleton_updated signal, but that is all.

**Reply by GuilhermeGSousa** [reply]
I just want to address some concerns I have with motion matching not being an animation node. There are a lot of cool features we get from using the animation tree we'd lose out on. Some examples:

Adding motion matching (filtering out certain bones) with another animation (like aiming)
![image](https://github.com/user-attachments/assets/da9ad162-d4dc-46e4-b859-7b6e3fb0f0f8)

Motion matching is more performant and accurate the more specialized the library that it runs on is. Since each MMAnimationNode only runs queries on a specific animation library, using nodes allows us to use state machines to do this
![image](https://github.com/user-attachments/assets/2133e152-a345-4eb4-9cef-af6c87373ce2)
where each state only runs on a specific animation library, giving us better animation selection.

> The most problematic issue here is the possibility of MotionMatching's next query input being played back in parallel, which must be avoided.

This is a very valid concern, but wouldn't it the same thing happen for a `AnimationNodeBlendSpace2D` that gets its `Vector2` from a script? Or a state machine that changes state based on some condition? In all these cases, these are nodes that are playing a specific animation based on animation tree parameters.

> MotionMatching would probably need to get the final pose of the skeleton with the Skeleton3d.skeleton_updated signal, but that is all.

I don't think this is the case: the data used to run motion matching does not need any information from IK computations. The initial pose of the skeleton at the beginning of the frame is sufficient to be able to run motion matching.

**Reply by GuilhermeGSousa** [reply]
I'll also add, in our current demo we are running motion matching without using any bone information at all, we're only basing selection off of trajectory predictions.

**Reply by TokageItLab** [reply]
What I was expecting from MotionMatching is to blend close animations from physics bone simulation results. So, in this MotionMatching solver, does it not make a distinction between climbing small steps and jumping? If so, then the conversation is quite simpler and the only thing missing is the struct API.

For example, if there is a pre-calculation with IK, then if the Skeleton's movement is upward and there is a difference in foot position, it is a small step and can be distinguished from a jump. Also, Skeleton's movement is forward and IK and Constraint make shrink the spine, the solver can determine that a crouch walk is necessary, but such a thing would require order resolution with the SkeletonModifier. Which is what I thought we were discussing, but it's not so, right?

**Reply by GuilhermeGSousa** [reply]
That distinction is generated by the trajectory, which does account for physics, which is, simply put, a simulation of what would happen if the character body kept moving. To take the examples you provided:

- If the player is climbing a hill, MMCharacter will generate a trajectory going up a hill. Motion matching then looks for animations with similar trajectories where the root bone goes up and play the one that matches best.
- If the player jumps, the future trajectory points will reflect the curvature of a jump. The same thing then happens on the motion matching side, we match the trajectory that fits that curvature best.

Bone positions/rotations are still useful for selection, but they mostly ensure that the animations being selected have similar skeletal poses, which can't be guaranteed when using trajectories alone.

**Reply by GuilhermeGSousa** [reply]
I think the only difference between what you and I are saying is that the physics based computations don't need to be done using IK on bones, and therefore aren't an input for motion matching. 

Going back to the initial discussion, we were addressing a way of doing inertialization blending, which happens after the motion matching node has run. That does try to have physically accurate bone movement when blending animations, but isn't used to make a motion matching query (we can run motion matching without it)

**Reply by TokageItLab** [reply]
If you are referring to the fact that in inertial blending, some transforms cannot be generated by interpolation of animation key values alone, then that is also probably beyond the role of AnimationNode/AnimationMixer.

AnimationNode only changes the time and the amount of blending. Also, the AnimationMixer only generates values by blending the animation key values.

Perhaps the hack of generating a dummy animation every frame using the AnimationProvider approach (https://github.com/Malcolmnixon/AnimationProviderDemo & https://github.com/Malcolmnixon/GodotXRAnimationRecorder) might be informative, but it raises performance concerns (And for this reason, finally XRAnimation was implemented as a SkeletonModifier).

**Reply by GuilhermeGSousa** [reply]
Generating a dummy animation is interesting, I will look into those implementations. The idea I initially had was:

- MMAnimationNode would switch from animation A to B without any blending.
- At the time of this switch, an inertial blending skeleton modifier would adjust bone position to maintain inertia and smoothly transition from animation A to B. The only thing preventing us from doing this right now is that the inertial blending skeleton modifier has no way of knowing when the MMAnimationNode switched animations (at least with the current API)

**Reply by momoluna444** [reply]
Maybe we could have an AnimationNode called “Cached Animation” that automatically caches the last N frames of the input subtree. Users can inherit this node and do whatever they want with the cached animation (motion matching, custom blending, IK, etc.). Since it's basically just a piece of temporary animation, AnimationMixer can mix it as usual. But I'm not sure how caching node subtree should be implemented, and the performance implications of editing animation frames at runtime. @TokageItLab what do you think?

**Reply by TokageItLab** [reply]
@tele4none As I have explained many times in many places, AnimationNode does not have values that can be applied directly to objects. It is the AnimationMixer or SkeletonModifier that applies values to objects, not the AnimationNode. So it is beyond the role of an AnimationNode.

All an AnimationNode can have is a reference to a specific Animation, which frame of that Animation to select, and the amount of blending.

In other words, as mentioned above, if you generate dummy animations in NodeAnimation, you can pass arbitrary values to AnimationMixer by make the AnimationNode refer that. However, editing the Animation key while the AnimationMixer is processing is not recommended currently from a safety or performance standpoint. Well, it would be possibly beneficial to consider an API specialized for dummy animations that solves these problems.

**Reply by momoluna444** [reply]
Yes, I see what you mean, and that's exactly how the idea came up while reading your previous comment.

Let me describe it in pseudocode:

NodeTimeInfo AnimationNodeCachedAnimation::_process(...) {
    tree->cache_input_to_animation(animation); // push current "Pose" to animation
    ...
    user_animation_modifier(animation);
    ...
    blend_animation(animation);
    return nti;
}

The caching part happens in AnimationTree, and the CachedAnimation node behaves like an AnimationNodeAnimation.

I also like the idea of AnimationMixerResultMixer. Anyway, it would be great to be able to blend and modify animations freely.

**Comment by pasqui23** [comment]
Speaking of the Inertilalization Blending, how will it interact with PhysicalSkeletonModifier? Will it allow implementing active ragdoll without instantiation the same skeleotn twice, first for the animation and 2nd for applying t porces onto it?

**Reply by GuilhermeGSousa** [reply]
I think that could be doable, assuming the order in which skeleton modifiers get applied. Imo support for inertialization blending needs to come from the engine, right now there's no good way for a skeleton modifier to know the velocity of a given bone, which is necessary to be able to do this.

**Reply by pasqui23** [reply]
To record the veplocity, couldn't you copy the position of every bone in the previous frame, compute the difference with the position in the current frame and divide it by delta?

**Reply by GuilhermeGSousa** [reply]
Knowing the velocity of the bone isn't enough. Its been a while since I've looked at it, but to implement inertialization you need to know the position of the bone before and after the animation get applied, and Skeleton Modifiers only have access to the latter (which they then modify).
You also need to know when an animation changes to be able to inertialize the transition, which get complicated when you use complex animation graphs. Imo there needs to be a way in Godot to extend blending behavior in the blend graph

**Comment by GeorgeS2019** [comment]
There is new upgrade in IK in Godot 4.6 and new physics demo on ragdoll in Godot demo..
Would be great to integrate to dream API
